kernel/src/grant.rs

```diff
@@ -343,7 +389,111 @@ impl KernelManagedLayout {
         // The location of the grant data T is the last element in the entire
         // grant region. Caller must verify that memory is accessible and well
         // aligned to T.
-        NonNull::new_unchecked(base_ptr.add(grant_size - grant_t_size.0))
+        let grant_t_size_usize: usize = grant_t_size.0;
+        NonNull::new_unchecked(base_ptr.add(grant_size - grant_t_size_usize))
+    }
+
+    /// Read an 8 bit value from the counter field offset by the specified
+    /// number of bits. This is a helper function for reading the counter field.
+    fn get_counter_offset(&self, offset_bits: usize) -> usize {
+        // # Safety
+        //
+        // Creating a `EnteredGrantKernelManagedLayout` object requires that the pointers
+        // are well aligned and point to valid memory.
+        let counters_val = unsafe { self.counters_ptr.read() };
+        (counters_val >> offset_bits) & 0xFF
+    }
+
+    /// Return the number of upcalls stored by the core kernel for this grant.
+    fn get_upcalls_number(&self) -> usize {
+        self.get_counter_offset(0)
```

Comment:
```suggestion
        u32::to_be_bytes(unsafe { self.counters_ptr.read() } as u32)[0]
```

Gets away without the helper function, same for `allow_rw/ro`.

Reply:
I thought I commented here, but using a helper function allows only one unsafe and not having to copy the safety comment several times.